src/mcpwm/comparator.rs

```diff
@@ -0,0 +1,127 @@
+use core::ptr;
+
+use esp_idf_sys::{
+    esp, mcpwm_cmpr_handle_t, mcpwm_comparator_config_t, mcpwm_comparator_config_t__bindgen_ty_1,
+    mcpwm_gen_compare_event_action_t, mcpwm_gen_handle_t, mcpwm_gen_t, mcpwm_new_comparator,
+    mcpwm_oper_handle_t, mcpwm_timer_direction_t_MCPWM_TIMER_DIRECTION_DOWN,
+    mcpwm_timer_direction_t_MCPWM_TIMER_DIRECTION_UP,
+};
+
+use super::generator::{CountingDirection, NoCmpMatchConfig, OnMatchCfg};
+
+trait ComparatorChannel {}
+
+pub struct CmpX;
+impl ComparatorChannel for CmpX {}
+
+pub struct CmpY;
+impl ComparatorChannel for CmpY {}
+
+pub trait OptionalCmp {
+    type OnMatchCfg: OnMatchCfg;
+    type Cfg: OptionalCmpCfg<Cmp = Self>;
+
+    fn get_comparator_mut(&mut self) -> Option<&mut Comparator>;
+}
+
+impl OptionalCmp for Comparator {
+    type OnMatchCfg = super::generator::CountingDirection;
+    type Cfg = ComparatorConfig;
+
+    fn get_comparator_mut(&mut self) -> Option<&mut Comparator> {
+        Some(self)
+    }
+}
+
+pub struct NoCmp;
+impl OptionalCmp for NoCmp {
+    type OnMatchCfg = super::generator::NoCmpMatchConfig;
+    type Cfg = NoCmpCfg;
+
+    fn get_comparator_mut(&mut self) -> Option<&mut Comparator> {
+        None
+    }
+}
+
+pub struct Comparator(pub(crate) mcpwm_cmpr_handle_t);
+
+impl Comparator {
+    pub(crate) unsafe fn configure(&mut self, gen: &mut mcpwm_gen_t, cfg: CountingDirection) {
+        extern "C" {
+            fn mcpwm_generator_set_actions_on_compare_event(
+                gen: mcpwm_gen_handle_t,
+                ev_act0: mcpwm_gen_compare_event_action_t,
+                ev_act1: mcpwm_gen_compare_event_action_t,
+                ev_act_end: mcpwm_gen_compare_event_action_t,
+            ) -> esp_idf_sys::esp_err_t;
+        }
+
+        esp!(mcpwm_generator_set_actions_on_compare_event(
+            gen,
+            mcpwm_gen_compare_event_action_t {
+                direction: mcpwm_timer_direction_t_MCPWM_TIMER_DIRECTION_UP,
+                comparator: self.0,
+                action: cfg.counting_up.into(),
+            },
+            mcpwm_gen_compare_event_action_t {
+                direction: mcpwm_timer_direction_t_MCPWM_TIMER_DIRECTION_DOWN,
+                comparator: self.0,
+                action: cfg.counting_down.into(),
+            },
+            mcpwm_gen_compare_event_action_t {
+                // <-- This marks the last argument in the variadic list
+                comparator: ptr::null_mut(),
+                ..Default::default()
+            }
+        ))
+        .unwrap();
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct ComparatorConfig {
+    flags: mcpwm_comparator_config_t__bindgen_ty_1,
+}
+
+impl Default for ComparatorConfig {
+    fn default() -> Self {
+        let mut flags: mcpwm_comparator_config_t__bindgen_ty_1 = Default::default();
+        // TODO: What should be set here?
+        flags.set_update_cmp_on_tep(0);
+        flags.set_update_cmp_on_tez(1);
```

Comment:
Just remembered https://github.com/espressif/esp-idf/issues/9904 . Do you think `tep` be set to `1` to avoid that? I have not tested this PR for that problem yet

Reply:
```suggestion
        flags.set_update_cmp_on_tep(1);
        flags.set_update_cmp_on_tez(1);
```